torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -692,6 +694,43 @@ def next_iter(self):
                 self.warn_status = _ExecOrderWarnStatus.WARNED
 
 
+class FreeEventQueue:
+    """
+    This tracks all pending frees corresponding to in-flight all-gathers. The
+    queueing pattern is iterative enqueues followed by a flush, and the current
+    heuristic for the flush is based on the number of inflight all-gathers.
+    """
+
+    def __init__(self) -> None:
+        self._queue: Deque[torch.cuda.Event] = collections.deque()
+        self._max_num_inflight_all_gathers = 2  # empirically chosen default
+
+    def enqueue(self, free_event: torch.cuda.Event) -> None:
+        """Enqueues a free event."""
+        self._queue.append(free_event)
+
+    def flush_if_needed(self) -> List[torch.cuda.Event]:
+        """
+        If the queue should be flushed (based on an internal criteria), then
+        this returns a non-empty :class:`list` of free events. Otherwise, this
+        returns an empty :class:`list`.
+        """
+        events: List[torch.cuda.Event] = []
+        if len(self._queue) >= self._max_num_inflight_all_gathers:
+            while self._queue:
+                event = self._dequeue()
+                assert event is not None
+                events.append(event)
+        return events
+
+    def _dequeue(self) -> Optional[torch.cuda.Event]:
+        """Dequeues a free event if possible."""
+        if self._queue:
+            event = self._queue.popleft()
+            return event
```

Comment:
nit: just `return self._queue.popleft()`?

Reply:
It appears that `popleft()` cannot take a default argument to return if the queue is empty, and instead, it will raise `IndexError` if the queue is empty. In that case, I believe we need to check non-emptiness before calling `popleft()`.